test/benchmark.py

```diff
@@ -0,0 +1,263 @@
+
+
+class SQLAlchemyBenchmarkTest(BenchmarkTestBase):
+    """Spanner dialect for SQLAlchemy performance testing class."""
+
+    def __init__(self):
+        super().__init__()
+        self._engine = create_engine(
+            "spanner:///projects/appdev-soda-spanner-staging/instances/"
+            "sqlalchemy-dialect-test/databases/compliance-test"
+        )
+        metadata = MetaData(bind=self._engine)
+        self._table = Table("Singers", metadata, autoload=True)
+
+        self._conn = self._engine.connect()
+
+        self._many_rows = []
+        num = 1
+        birth_date = datetime.datetime(1998, 10, 6).strftime("%Y-%m-%d")
+        for i in range(99):
+            num += 1
+            self._many_rows.append(
+                {
+                    "id": num,
+                    "first_name": "Pete",
+                    "last_name": "Allison",
+                    "birth_date": birth_date,
+                    "picture": b"123",
+                }
+            )
+
+    @measure_execution_time
+    def insert_one_row_with_fetch_after(self):
+        self._conn.execute(insert(self._table).values(self._one_row))
+        last_name = self._conn.execute(
+            select([text("last_name")], from_obj=self._table)
+        ).fetchone()[0]
+        if last_name != "Allison":
+            raise ValueError("Received invalid last name: " + last_name)
+
+    @measure_execution_time
+    def insert_many_rows(self):
+        self._conn.execute(
+            self._table.insert(), self._many_rows,
+        )
+
+    @measure_execution_time
+    def read_one_row(self):
+        row = self._conn.execute(select(["*"], from_obj=self._table)).fetchone()
+        if not row:
+            raise ValueError("No rows read")
+
+    @measure_execution_time
+    def select_many_rows(self):
+        rows = self._conn.execute(select(["*"], from_obj=self._table)).fetchall()
+        if len(rows) != 100:
+            raise ValueError("Wrong number of rows read")
+
+
+def insert_one_row(transaction, one_row):
+    """A transaction-function for the original Spanner client.
+
+    Inserts a single row into a database and then fetches it back.
+    """
+    transaction.execute_update(
+        "INSERT Singers (id, first_name, last_name, birth_date, picture) "
+        " VALUES {}".format(str(one_row))
+    )
+    last_name = transaction.execute_sql(
+        "SELECT last_name FROM Singers WHERE id=1"
+    ).one()[0]
+    if last_name != "Allison":
+        raise ValueError("Received invalid last name: " + last_name)
+
+
+def insert_many_rows(transaction, many_rows):
+    """A transaction-function for the original Spanner client.
+
+    Insert 100 rows into a database.
+    """
+    total_count = 0
+    for row in many_rows:
+        count = transaction.execute_update(
+            "INSERT Singers (id, first_name, last_name, birth_date, picture) "
+            " VALUES {}".format(str(row))
+        )
+        total_count += count
```

Comment:
I think if this was changed to using the [batch mutations API](https://cloud.google.com/spanner/docs/getting-started/python#write-data-with-mutations), then there might be a noticeable difference between the spanner client and SQLAlchemy. Can you also run a test for that to see what the difference is? Because we will have to support it for GA even if we don't for preview.

Reply:
Yeah, that's true, here are first measures:
```python
{'insert_many_rows_with_mutations': {
    'SQLAlchemy deviation': 33.33,
    'SQLAlchemy to Spanner, %': 2644,
    'SQLAlchemy, sec': 34.64,
    'Spanner, sec': 1.31
}}
```
Hm-m, I'll take a look if we can override something somewhere to make SQLAlchemy execute inserts with mutations instead of `Cursor.execute_many()`, but no promises yet.